OmniAccount auth rework

silva-fj opened the pull request:
In this PR we are refactoring the OmniAccount authentication mechanism without changing it's behaviour. The changes consist of simplifying the data structures and separation of concerns (putting together auth related data in `OmniAuth`.

It includes the following sub-tasks:
- [X]  https://linear.app/litentry/issue/P-1453/re-design-and-implement-the-msg-for-web3-authentication
- [X]  https://linear.app/litentry/issue/P-1454/include-all-information-in-omniauth
- [X] https://linear.app/litentry/issue/P-1456/jsoned-auth-struct

**Note:** The new json structs for authentication (including `Identity` and `HeimaMultiSignature`) have been included, but the params of the RPC methods have not been changed as we have some inconsistencies in the way we handle authentication, mostly because of some pumpx specific logic. We should refactor the params once we define the (generic) API we want to expose for other potential clients in a follow up PR.

silva-fj committed:
feat(auth): implement Web3 authentication message codes

Add time-based authentication message code generator for Web3
authentication flows.

silva-fj committed:
refactor(auth): simplify Web3 authentication flow

Replace complex nonce-based authentication with a time-based message
code system for Web3 authentication.

silva-fj committed:
test(auth): add unit test for Web3 authentication

silva-fj committed:
feat(rpc): add omni_getMessageCode RPC method

Introduce the `omni_getMessageCode` RPC method to generate and return a
message code for web3 authentication.

silva-fj committed:
feat(web3-auth): use message code for task signing

Refactored integration tests and type creators to use a dynamic message
code for signing native tasks in Web3 authentication flows.

silva-fj committed:
refactor(rpc): remove mrenclave from server context

silva-fj committed:
fixing typo

silva-fj committed:
Merge branch 'dev' into p-1453-re-design-and-implement-the-msg-for-web3-authentication

silva-fj committed:
Update the `verify_auth` function to accept explicit auth and
sender parameters, removing the need to pass the entire wrapper.

silva-fj committed:
feat(jwt): require "sub" and "typ" claims in JWT validation

silva-fj committed:
update auth_token validation

silva-fj committed:
feat(auth): refactor JWT validation and usage in pumpx rpc methods

Refactor JWT validation in pumpx RPC methods to use the new
verify_auth_token_authentication function, which enforces the presence
of the "typ" claim and returns claims on success. Remove redundant key
handling and direct JWT decoding. Update submit_swap_order to check
user identity against the token subject.

silva-fj committed:
refactor: improve web3 msg code testability

silva-fj committed:
Merge branch 'dev' into p-1453-re-design-and-implement-the-msg-for-web3-authentication

silva-fj committed:
feat(auth): return seconds left with message code

Update the message code generation to return both the code and the number
of seconds left in the current period. Adjust the RPC response and
verification logic to accommodate this change. Add tests for period
boundary behavior and seconds left calculation.

silva-fj committed:
updating integration tests

silva-fj committed:
revert changes to pumpx rpc methods

silva-fj committed:
feat(auth): validate JWT subject matches user identity

silva-fj committed:
Merge branch 'dev' into p-1453-re-design-and-implement-the-msg-for-web3-authentication

silva-fj committed:
feat(core): add Hashable trait and impl for AccountId

silva-fj committed:
feat(helpers): make generate_otp function public

silva-fj committed:
feat(rpc): redesign web3 message code generation and verify

Refactor the omni_getMessageCode RPC to generate and store a unique OTP
message code per omni account, replacing the previous static code approach.
Update web3 authentication verification to fetch and validate the stored code
from persistent storage. Unify verification error handling and update tests
to reflect the new logic.

silva-fj committed:
test(omni-account): update getMessageCode usage in tests

Refactor integration tests to pass the omniAccount parameter to
getMessageCode, aligning with the updated function signature. This ensures
tests correctly request message codes for specific omni accounts and match
the new RPC API requirements.

silva-fj committed:
refactor(auth): remove chrono dependency and time-based code

Eliminate the unused chrono dependency and related time-based message code
generation logic from the authentication module.

silva-fj committed:
Merge branch 'dev' into p-1453-re-design-and-implement-the-msg-for-web3-authentication

silva-fj committed:
Merge branch 'p-1453-re-design-and-implement-the-msg-for-web3-authentication' into p-1479-check-the-typ-of-incoming-jwt-token

silva-fj committed:
refactor(auth): unify omni account hex usage in verification

Update authentication verification to consistently use the hex-encoded omni
account string instead of Identity references.

silva-fj committed:
fixing dev-dependencies

silva-fj committed:
Merge branch 'p-1453-re-design-and-implement-the-msg-for-web3-authentication' into p-1479-check-the-typ-of-incoming-jwt-token

silva-fj committed:
Merge remote-tracking branch 'origin/dev' into p-1453-re-design-and-implement-the-msg-for-web3-authentication

silva-fj committed:
Merge branch 'dev' into p-1453-re-design-and-implement-the-msg-for-web3-authentication

silva-fj committed:
fixing merging issues

silva-fj committed:
Merge branch 'dev' into p-1453-re-design-and-implement-the-msg-for-web3-authentication

silva-fj committed:
Merge remote-tracking branch 'origin/p-1453-re-design-and-implement-the-msg-for-web3-authentication' into p-1479-check-the-typ-of-incoming-jwt-token

silva-fj committed:
feat(auth): updating OmniAuth to include sender's data

Refactor OmniAuth enum to include sender identity in AuthToken and Web3
variants. Update RPC methods to construct AuthToken with sender's
omni_account hex and JWT token.

silva-fj committed:
fix verify_auth_token_authentication call on omni/submit_swap_order

silva-fj committed:
feat(auth): include sender info in OmniAuth variants

Refactor OmniAuth::AuthToken and Web3 variants to include sender
information. Update all RPC methods and authentication verification
logic to use the new OmniAuth structure.

BillyWooo committed:
Merge branch 'dev' into p-1452-rework-the-auth-of-omni-account

silva-fj committed:
Merge branch 'dev' into p-1452-rework-the-auth-of-omni-account

silva-fj committed:
feat(auth): add serde support for Identity and signatures

Introduce serializable representations for `Identity` and `HeimaMultiSignature`
enabling JSON interchange between client and server. Implement conversion
from these serde types to their internal Rust counterparts

silva-fj committed:
updating comments

silva-fj committed:
fix(omni): improve error logging in native task parsing

Add detailed error logs for failures during native task parsing,
decryption, and authentication in the submit_native_task method. This
enhances traceability and debugging of request handling issues.

silva-fj committed:
adding logs

silva-fj committed:
updating type definitions

silva-fj committed:
updating createNativeTaskWrapper

silva-fj committed:
Merge branch 'dev' into p-1452-rework-the-auth-of-omni-account

kziemianek commented on tee-worker/omni-executor/rpc-server/src/methods/omni/get_message_code.rs:
```diff
@@ -0,0 +1,34 @@
+			let message_code = generate_otp(8);
+
+			verification_code_storage
+				.insert(&omni_account.hash(), message_code.clone())
```
My two concerns here:
* this is an only growing storage (old codes are not pruned - can be only overriden - it's relatively cheap for somebody to flood worker with prepared requests  - I haven't measured the impact)
* people can override code for each other

silva-fj replied:
> this is an only growing storage (old codes are not pruned - can be only overriden - it's relatively cheap for somebody to flood worker with prepared requests - I haven't measured the impact)

This is a good point. I'll remove the entry after the code is verified.

> people can override code for each other

Should we make this idempotent until the code is used? If there is a code in storage we just return it instead of inserting a new one?

kziemianek replied:
> This is a good point. I'll remove the entry after the code is verified.

Still unused codes remain in storage - if that could be an issue maybe we should have a cron job for cleaning them up

> Should we make this idempotent until the code is used? If there is a code in storage we just return it instead of inserting a new one?

perfect 🙃

silva-fj replied:
> Still unused codes remain in storage - if that could be an issue maybe we should have a cron job for cleaning them up

Yeah it would be a good improvement

silva-fj committed:
Merge remote-tracking branch 'origin/dev' into p-1452-rework-the-auth-of-omni-account

silva-fj committed:
removing legacy definitions

silva-fj committed:
fix(auth): remove verification code after retrieval

Ensure the verification code is removed from storage after it is retrieved
for authentication. This prevents reuse of the same code and improves
security by enforcing one-time usage of verification codes.

silva-fj committed:
fix(auth): reuse verification code if it exists

Update get_message_code to check for an existing verification code before
generating a new one. If a code exists for the omni account, it is reused;
otherwise, a new code is generated and stored. This prevents overwriting
active codes and ensures consistent authentication behavior.

kziemianek commented on tee-worker/client-api/src/omni/interfaces/omniExecutor/definitions.ts:
```diff
@@ -41,10 +41,10 @@ export default {
       _enum: {
-        Web3: "(HeimaMultiSignature)",
+        Web3: "(Identity, HeimaMultiSignature)",
         Email: "(Text, Text)",
```
Can all variants contain `Identity` as a first value ?

silva-fj replied:
They could, one thing is a bit annoying is that we'll have to add conversions/checks for the identity. But we can revisit this when we actually start using the serde types (so the clients send json objects) instead of these internal types

kziemianek commented on tee-worker/omni-executor/executor-primitives/src/auth.rs:
```diff
@@ -1,36 +1,217 @@
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(tag = "type", content = "data")]
+pub enum HeimaMultiSignatureSerde {
+	Ed25519(String),  // base64-encoded
```
I would suggest keeping hex encoding here.
These are short length byte arrays and I see no benefit from using base64, do you?

kziemianek commented on tee-worker/omni-executor/executor-primitives/src/auth.rs:
```diff
@@ -1,36 +1,217 @@
+///   "data": "base64-encoded-signature"
+/// }
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
+#[serde(tag = "type", content = "data")]
```
If you move this serde macro to `HeimaMultiSignature` and enable `sp-core` `serde` feature then you can json-serialize `HeimaMultiSignature` directly to:

```
{"type":"Ed25519","data":"3e1994ba3589f8ae95bbe118ba30186d641b95c44205de8c161088ef9a1685604f02b46a9670740b0623200491f03682cec1c839f170237ae261aee7dd0d6202"}
```

If we can stick to hex encoding maybe `HeimaMultiSignatureSerde` is not needed at all?

silva-fj replied:
Done ✅

kziemianek reviewed (approved):
It looks fine. I've just left some cosmetic comments.

silva-fj committed:
Merge remote-tracking branch 'origin/dev' into p-1452-rework-the-auth-of-omni-account

silva-fj committed:
Merge branch 'dev' into p-1452-rework-the-auth-of-omni-account

silva-fj committed:
feat(auth): switch multi-signature serde to hex encoding

silva-fj committed:
feat(auth): derive serde for HeimaMultiSignature

Move serde serialization and deserialization for HeimaMultiSignature
directly into the enum using serde attributes. Remove the
HeimaMultiSignatureSerde wrapper and update all usages and tests to use
the enum directly. This simplifies the codebase and ensures consistent
serialization for multi-signature types.

silva-fj committed:
fixing clippy issues

silva-fj committed:
Merge branch 'dev' into p-1452-rework-the-auth-of-omni-account

silva-fj committed:
Merge branch 'dev' into p-1452-rework-the-auth-of-omni-account
